File: basemodule/src/main/java/com/test/basemodule/base/view/adapter/BaseViewHolder.java
package com.test.basemodule.base.view.adapter;

import android.content.Context;
import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

/**
 * Created by Michael on 12/25/17.
 */

public abstract class BaseViewHolder<E> extends RecyclerView.ViewHolder{
    protected OnItemClickListener<E> onItemClickListener;
    protected E itemData;

    public BaseViewHolder(View itemView) {
        super(itemView);
        itemView.setOnClickListener(v -> {
            if(onItemClickListener != null)
                onItemClickListener.onItemClick(v, itemData);
        });
    }


    public void setOnItemClickListener(OnItemClickListener<E> onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void bind(E itemData){
        this.itemData = itemData;
        onBindView(itemData);
    }

    protected abstract void onBindView(E itemData);

    protected Context getContext()
    {
        return itemView.getContext();
    }

    protected  E getItemData()
    {
        return itemData;
    }
}

File: basemodule/src/main/java/com/test/basemodule/base/view/adapter/BaseViewHolder.kt
package com.test.basemodule.base.view.adapter

import android.content.Context
import android.view.View
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by Michael on 12/25/17.
 */
abstract class BaseViewHolder<DataType, Binding : ViewDataBinding>(protected val binding: Binding) :
    RecyclerView.ViewHolder(binding.root) {
    var onItemClickListener: OnItemClickListener<DataType>? = null
    protected var itemData: DataType? = null

    init {
        itemView.setOnClickListener { v: View? ->
            if (onItemClickListener != null)
                onItemClickListener!!.onItemClick(v, itemData)
        }
    }

    fun bind(itemData: DataType) {
        this.itemData = itemData
        onBindView(itemData)
    }

    protected abstract fun onBindView(itemData: DataType)

    protected val context: Context
        get() = itemView.context


}